src/sagas/wallet.js

```diff
@@ -386,24 +387,71 @@ export function* handleTx(action) {
   }
 
   // find tokens affected by the transaction
-  const balances = yield call(wallet.getTxBalance.bind(wallet), tx);
   const stateTokens = yield select((state) => state.tokens);
   const registeredTokens = stateTokens.map((token) => token.uid);
 
-  // We should download the **balance** for every token involved in the transaction
-  // and history for hathor and DEFAULT_TOKEN
-  for (const [tokenUid] of Object.entries(balances)) {
-    if (registeredTokens.indexOf(tokenUid) === -1) {
-      continue;
+  // To be able to only download balances for tokens belonging to this wallet, we
+  // need a list of tokens and addresses involved in the transaction from both the
+  // inputs and outputs.
+  const { inputs, outputs } = tx;
+  const data = [...inputs, ...outputs];
+
+  // tokenAddressesMap should be an object { [tokenUid]: Set(addresses) }
+  // txAddresses should be a Set with all addresses involved in the tx
+  const [tokenAddressesMap, txAddresses] = data.reduce(
+    (acc, io) => {
+      if (!io.decoded) {
+        return acc;
+      }
+
+      const { token, decoded: { address } } = io;
+
+      // We are only interested in registered tokens
+      if (registeredTokens.indexOf(token) === -1) {
+        return acc;
+      }
+
+      if (!acc[token]) {
+        acc[0][token] = new Set([]);
+      }
+
+      acc[0][token].add(address);
+      acc[1].add(address);
+
+      return acc;
+    }, [{}, new Set([])],
+  );
+
+  const txWalletAddresses = yield call(wallet.checkAddressesMine.bind(wallet), [...txAddresses]);
+  const tokensToDownload = [];
+
+  for (const [tokenUid, addresses] of Object.entries(tokenAddressesMap)) {
+    for (const [address] of addresses.entries()) {
+      // txWalletAddresses should always have the address we requested, but we should double check
+      // here using lodash just in case
+      const inWallet = get(txWalletAddresses, address, false);
```

Comment:
If we use `wallet.isAddressMine` as a filter when creating the `tokenAddressesMap` we can ensure that the keys of `tokenAddressesMap` only have tokens with at least one address in the wallet.

Reply:
I agree, but that would only work on the fullnode facade, the wallet-service facade makes a http request to the wallet-service lambda to detect if the address belongs to the wallet